### PR TITLE
OAK-11204: Incorrect order with more than 2^16 segments in remote archive

### PR DESCRIPTION
**Pull request by lweitzendorf**
Adds the ability for `SegmentArchiveWriter` implementations to communicate a limitation in segment count support number. This is used by `TarFiles` to roll over to a new writer if needed.

### CODE REVIEW

**Comment by mreutegg** [comment]
This looks useful.

Please create a [JIRA](https://issues.apache.org/jira/browse/OAK) ticket and follow guidelines described on https://jackrabbit.apache.org/oak/docs/participating.html#commits

More specifically, use the ticket id as a prefix for the commit message.

File: oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/file/tar/SegmentTarWriter.java [hunk 1]
```diff
@@ -93,7 +94,7 @@ public SegmentTarWriter(File file, FileStoreMonitor monitor, IOMonitor ioMonitor
     }
 
     @Override
-    public void writeSegment(long msb, long lsb, byte[] data, int offset, int size, int generation, int fullGeneration, boolean compacted) throws IOException {
+    public void writeSegment(long msb, long lsb, byte @NotNull [] data, int offset, int size, int generation, int fullGeneration, boolean compacted) throws IOException {
```
Comment (mreutegg) [inline oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/file/tar/SegmentTarWriter.java]: Please refrain from adding unrelated changes to the PR. The additional annotations are justified, but should be handled with a dedicated JIRA ticket / PR.

File: oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/file/tar/TarFilesTest.java [hunk 2]
```diff
@@ -445,4 +451,54 @@ public boolean shouldFollow(UUID from, UUID to) {
...
+                .withFileStoreMonitor(fsMonitor)
+                .withMaxFileSize(MAX_FILE_SIZE)
+                .withRemoteStoreMonitor(remoteStoreMonitor)
+                .withPersistence(persistence)
+                .build();
+
+        // write more entries than fit into a single tar file
+        for (int i = 0; i < TOTAL_ENTRIES; i++) {
+            writeSegment(randomUUID());
+        }
```
Comment (mreutegg) [inline oak-segment-tar/src/test/java/org/apache/jackrabbit/oak/segment/file/tar/TarFilesTest.java:502]: Would be good to assert that the number of created archives is as expected. I think `tarFiles.readerCount()` should be `5` at this point, right?

File: oak-segment-remote/src/main/java/org/apache/jackrabbit/oak/segment/remote/RemoteUtilities.java [hunk 3]
```diff
@@ -28,6 +28,7 @@
 public final class RemoteUtilities {
     public static final boolean OFF_HEAP = getBoolean("access.off.heap");
     public static final String SEGMENT_FILE_NAME_PATTERN = "^([0-9a-f]{4})\\.([0-9a-f-]+)$";
+    public static final int MAX_ENTRY_COUNT = 0x10000;
```
Comment (mreutegg) [inline oak-segment-remote/src/main/java/org/apache/jackrabbit/oak/segment/remote/RemoteUtilities.java:30]: I think this is off by one. `SEGMENT_FILE_NAME_PATTERN` only allows for 4 hex digits. Consider the following test.
```
    @Test
    public void maxEntryCount() {
        UUID uuid = UUID.randomUUID();
        String name = RemoteUtilities.getSegmentFileName(RemoteUtilities.MAX_ENTRY_COUNT, uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
        UUID parsed = RemoteUtilities.getSegmentUUID(name);
        assertEquals(uuid, parsed);
    }
```

Comment (lweitzendorf) [reply oak-segment-remote/src/main/java/org/apache/jackrabbit/oak/segment/remote/RemoteUtilities.java:30]: Indices start at 0, so with 0x10000 entries the highest index would be 0xffff, right?

Comment (mreutegg) [reply oak-segment-remote/src/main/java/org/apache/jackrabbit/oak/segment/remote/RemoteUtilities.java:30]: Ah, yes. You are right. RemoteUtilities.getSegmentFileName() takes a position (index) and not an entry count. The test should rather to the following.
```
    @Test
    public void maxEntryCount() {
        UUID uuid = UUID.randomUUID();
        String name = RemoteUtilities.getSegmentFileName(RemoteUtilities.MAX_ENTRY_COUNT - 1, uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
        UUID parsed = RemoteUtilities.getSegmentUUID(name);
        assertEquals(uuid, parsed);
    }
```

**Comment by mreutegg** [comment]
Build failure was in oak-store-document and looks unrelated to this PR.
